File: 3 Longest Substring Without Repeating Characters/soln.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size(), i = 0, j = 0, mx = 0;
        
        unordered_map<char, int> m;
        
        while (j < n){
            mx = max(mx, j - i);
            
            if (m.find(s[j]) != m.end()){
                if (m[s[j]] > 0){
                    while (m[s[j]] > 0){
                        m[s[i]] = m[s[i]] - 1;
                        i ++;
                    }
                }
            }
            if (m.find(s[j]) == m.end()){
                m[s[j]] = 1;
            } else{
                m[s[j]] = m[s[j]] + 1;
            }
            j ++;
        }
        mx = max(mx, j - i);
        return mx;
    }
};

File: 3 Longest Substring Without Repeating Characters/soln.go
func lengthOfLongestSubstring(s string) int {
    n := len(s)
    i := 0
    j := 0
    mx := 0
    m := make(map[string]int)
    
    for j < n{
        if mx < j - i{
            mx = j - i
        }
        if _, ok := m[string(s[j])]; ok{
            if m[string(s[j])] > 0{
                for m[string(s[j])] > 0{
                    m[string(s[i])] = m[string(s[i])] - 1
                    i = i + 1
                }
            }
        }
        m[string(s[j])] = m[string(s[j])] + 1
        j = j + 1
    }
    if mx > j - i{
            mx = mx
        } else{
            mx = j - i
        }
    return mx
}